text/0035-schema-applies-to-unspecified.md

```diff
@@ -0,0 +1,122 @@
+Your current schema will result in error when you upgrade to Y.
+If you use your modified schema with X, you will be able to validate policies that will not validate in Y.
+There is _no way_ to write a schema so that validation behavior will be the same between X and Y.
+
+In **case 2**, you should update your schema to use the following syntax instead.
+
+```json
+"appliesTo": {
+    "context": {}
+}
+```
+
+This will have the same meaning in both X and Y: the action applies to an unspecified principal and resource, and the empty context.
+
+## Drawbacks
+
+This is a breaking change that will alter the meaning of current schemas.
+See the discussion on backwards compatibility [above](#notes-on-backwards-compatibility) and discussion on backwards-compatible alternatives [below](#alternatives).
+
+## Alternatives
+
+### Alternative A
+
+The main downside of the current proposal is that, in some cases (see **case 1** [above](#notes-on-backwards-compatibility)), there will be no way to update your schema before you upgrade to Cedar Y -- you will have to update your schema simultaneously with the upgrade.
+This will be a problem for large applications where updating policies, schemas, and the Cedar SDK cannot be done atomically.
+
+We could mitigate this by instead implementing the following changes.
+
+| Scenario | Status Quo | Proposal |
+| --------- | -------- | ------- |
+| Specified `appliesTo` but omitted `principalTypes`/`resourceTypes` | Authz request with this action can only have an unspecified entity as the principal/resource | Same as status quo |
+| `appliesTo.principalTypes: []` AND `appliesTo.resourceTypes: []` | Action cannot be used in authz request | Same as status quo |
+| `appliesTo.principalTypes: []` OR `appliesTo.resourceTypes: []` | Action cannot be used in authz request | ❌ Not allowed |
+| Omitted `appliesTo` | Authz request with this action can only have unspecified entities as the principal and resource | ❌ Not allowed |
+
+These changes remove some of the confusing behavior of the current implementation (e.g., setting only one of `appliesTo.principalTypes`/`appliesTo.resourceTypes` to be the empty list), while maintaining backwards compatibility.
+Under this alternative, some schemas will produce parse errors after upgrading to Y, but users can fix these error proactively.
+For every valid schema in Y, there is some schema in X that has identical behavior.
+
+The downside of this alternative is that it still allows the empty list to mean that an action applies to no entities, which may be confusing for users.
+Specifying that an action applies to no entities also requires more characters with this alternative, possibly making it more annoying to type out.
+
+This alternative will require updates to [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md) in order to maintain consistency between the two schema formats.
```

Comment:
Original discussion: https://github.com/cedar-policy/rfcs/pull/35#discussion_r1380203955

Reply:
After writing this up, I actually like alternative A a little more that the original proposal because it alleviates backwards compatibility issues. Between these two options:
```json
"action": {}
```
```json
"action": {
  "appliesTo": {
    "principalTypes": [],
    "resourceTypes": []
  }
}
```
I think that both adequately convey that `"action"` applies to no principals or resources.

Reply:
I agree that both convey that `"action"` applies to no principals or resources. 

But this alternative has two downsides:
- It requires that every action group is specified more verbosely. We have at least one customer that makes heavy use of action groups, and this proposal makes schemas less readable for this customer. 
- It does match the custom schema syntax, which does allow omitting `appliesTo` (precisely so that schemas are less verbose). 

I suppose we could accept this proposal and then in the next major version to relax the prohibition on omitted `appliesTo`, so that it matches the custom syntax. This would amount to a breaking change followed by a backward compatible change, so perhaps that's not so bad.

But I would prefer we get some data about whether anyone is currently using the omitted `appliesTo` to mean an action that applies to unspecified principals and resources. I have a feeling that this is not actually used, in which case we are accepting a weaker alternative to no good purpose.